Update Configuration Component

MLopezJ opened the pull request:
More info: #256

MLopezJ committed:
feat: add 'configuration' constraints
constraints:
1- 'Movement resolution' must be higher than 'Accelerometer inactivity timeout'
2- 'Accelerometer activity threshold' must be higher than 'Accelerometer inactivity threshold'

if constraints are not satisfied, 'configuration' values can not be updated.

MLopezJ committed:
test: add 'components' folder route to test name mapper

MLopezJ committed:
fix: follow standard of capital letters on config keys

MLopezJ committed:
fix: reorder 'configuration' items

MLopezJ committed:
feat: add preset config value
Implement 2 present config values:
- Walking: when you want to track the movement of a person or a pet
- Parcel: When you want to track the movement of a parcel

The preset configs are the recommended values for each case.

MLopezJ committed:
fix: rename method

from 'configConstraints' to 'validateConfig'

MLopezJ committed:
fix: improve method import format

MLopezJ committed:
feat: add preset configs

Implement 2 preset configs value:
- Walking: when you want to track the movement of a person or a pet
- Parcel: When you want to track the movement of a parcel

Preset configs are the recommended values for each case.

MLopezJ committed:
refactor: simplify validateConfig tests

MLopezJ committed:
feat: create config explainer
Create a section to explain in text the meaning of the config inputs. this way is easier for the user to undertand the impact of their changes on the form

MLopezJ committed:
fix: set proper types

MLopezJ committed:
feat: redirect input field from 'config explainer'
Focus custom input field from the 'config explainer' component. This way, the user will be able to see the correlation between the explain section and the actual value

MLopezJ committed:
fix: remove useFocus hook
useFocus used to create a ref of a html element to be able to set the focus from SettingExplainer component.

The same behavior was addressed using useRef hook in the Setting component, this way useFocus is not needed anymore and for that reason was removed.
Update test as well

MLopezJ committed:
fix: rename prop
'ref' is not a valid prop name for a component

MLopezJ committed:
fix: remove cols
remove divs with className 'col', because the style that class give to component is not the required in ticket description

MLopezJ committed:
fix: add style to 'configuration explainer' section

MLopezJ committed:
fix: add style to 'preset configuration' section

MLopezJ committed:
fix: remove margin top from 'config explainer' section

MLopezJ committed:
refactor: e2e test for 'explainer configuration' section

MLopezJ committed:
refactor: e2e tests for 'Asset Configuration' constraints

MLopezJ committed:
refactor: e2e test for preset configurations

MLopezJ committed:
refactor: e2e test for 'Asset Settings' form

Commit:
style: auto-format source code

MLopezJ committed:
fix: update 'Asset Config' inputs order

MLopezJ committed:
fix: update 'Asset Config' inputs order

MLopezJ committed:
refactor: 'Presets' component
centralize the concerns of the component

Commit:
style: auto-format source code

MLopezJ committed:
fix: make e2e configuration constraints tests more legible

MLopezJ committed:
fix: make e2e configuration explainer tests more legible

Commit:
style: auto-format source code

MLopezJ committed:
docs: add TODO on e2e test

MLopezJ committed:
fix: rename css class on Presets component

MLopezJ committed:
fix: remove console.log()

MLopezJ committed:
fix: remove comment

MLopezJ committed:
test: presets

MLopezJ committed:
fix: rename files
uses 'settings' instead of 'configuration' to name files and folder, because the term 'settings' was used before on the project to refer to that component of the project

MLopezJ committed:
fix: update recommended settings explanation

MLopezJ committed:
fix: update 'Presets' component
Update UI for rational use of space reasons

Commit:
style: auto-format source code

coderbyheart commented on src/components/Asset/Settings/textAsButton.tsx:
```diff
@@ -0,0 +1,27 @@
+	className?: string
+}) => {
+	return (
+		<u
```
Use `button.link`

MLopezJ replied:
Can you elaborate more? 

I think I'm not getting the idea.

MLopezJ replied:
I mean, I'm not sure if this is requesting to use <a> tag with href prop or to update the TextAsButton component or something else.

coderbyheart commented on src/components/Asset/Settings/Settings.module.css:
```diff
@@ -65,8 +65,28 @@
   flex-direction: column;
 }
+
+.collap {
```
```suggestion
.collapse {
```

coderbyheart commented on src/components/Asset/Settings/Settings.module.css:
```diff
@@ -65,8 +65,28 @@
+  border-radius: 5px;
+}
+
+#rowOne {
```
Use selectors, not IDs

coderbyheart commented on src/components/Asset/Settings/Presets.tsx:
```diff
@@ -0,0 +1,67 @@
+				? presetConfigs[`${id}`].config
+				: currentDesiredConfig
+		setDesiredConfig(config)
+	}
```
No need to check here, because we have a valid config in the loop in line 40

coderbyheart commented on src/components/Asset/Settings/Presets.spec.tsx:
```diff
@@ -0,0 +1,78 @@
+			/>,
+		)
+
+		const parcel = presets.findOne('#walking-info')
```
Finde elements for tests using test attributes: https://playwright.dev/docs/selectors#define-explicit-contract

coderbyheart commented on src/components/Asset/Settings/Presets.module.css:
```diff
@@ -0,0 +1,32 @@
+  flex-direction: column;
+}
+
+.collap {
```
```suggestion
.collapse {
```

coderbyheart commented on src/components/Asset/Settings/NumberConfigSetting.tsx:
```diff
@@ -65,7 +74,9 @@ export const NumberConfigSetting = ({
 				/>
 				<input
-					className="form-control"
+					className={`form-control ${
```
Use `classNames`

coderbyheart commented on src/components/Asset/Settings/NumberConfigSetting.tsx:
```diff
@@ -26,10 +28,17 @@ export const NumberConfigSetting = ({
+
+	useEffect(() => {
+		updateInput(`${desired ?? reported}`)
+	}, [desired, reported, updateInput])
```
Why?

coderbyheart commented on src/asset/asset.ts:
```diff
@@ -47,7 +47,7 @@ export const AssetConfig = Type.Object(
 		accath: Type.Number({
 			description:
-				'Accelerometer activity threshold in m/s²: Minimal absolute value for an accelerometer reading to be considered movement.',
+				'Accelerometer Activity Threshold in m/s²: Minimal absolute value for an accelerometer reading to be considered movement.',
```
Make consistent

coderbyheart reviewed (commented):
Adding a fews things to go over together.

coderbyheart commented on src/asset/config.ts:
```diff
@@ -12,3 +12,43 @@ export const defaultConfig: Static<typeof AssetConfig> = {
+			nod: [],
+		},
+		label: 'Parcel',
+		description: 'Used for tracking parcels.',
```
Add more extensive description (explain what values are specifically set and why).

coderbyheart commented on src/asset/config.ts:
```diff
@@ -12,3 +12,43 @@ export const defaultConfig: Static<typeof AssetConfig> = {
+	}
+> = {
+	parcel: {
+		config: {
```
Only make it set values that are needed for this use case

MLopezJ committed:
fix: rename style class
from 'collap' to 'collapse'

MLopezJ committed:
fix: remove not needed validation
Valid config type already happening on line 32 because config has 'AssetConfig' type declaration

coderbyheart committed:
fix: use key to rerender number inputs if preset is used

coderbyheart commented on src/components/Asset/Settings/Presets.module.css:
```diff
@@ -0,0 +1,32 @@
+  flex-direction: column;
+}
+
+.collapse {
```
Make re-usable (it is used twice)

coderbyheart commented on src/components/Asset/Settings/NumberConfigSetting.tsx:
```diff
@@ -82,6 +90,7 @@ export const NumberConfigSetting = ({
 					}}
 				/>
 				<span className="input-group-text">{unit ?? 's'}</span>
+				<div className="invalid-feedback">{errorMessage}</div>
```
Do not render if no error message

coderbyheart committed:
test(e2e): use data property to find elements

MLopezJ committed:
fix: use classnames
use classnames instead of classname

MLopezJ committed:
test(e2e): use data property to find elements

MLopezJ committed:
fix: remove comment

MLopezJ committed:
fix: render div only when is required
add validation to render div related to invalid input only when there is an error message

MLopezJ commented on src/components/Asset/Personalization.tsx:
```diff
@@ -26,6 +26,7 @@ export const Personalization = ({ asset }: { asset: Asset }) => {
 					style={{ width: '100%' }}
 				>
 					<div className="flex-grow-1">
+						{/** TODO: chango to 'classNames' for consisten */}
```
I notice we made the same behavior we are requesting to change (https://github.com/NordicSemiconductor/asset-tracker-cloud-app-aws-js/pull/260#discussion_r977623543) here as well, but this component is not related to this PR at all. 

Should we update this here as well (for for consistency reasons) or should we just leave it as it is?

coderbyheart replied:
Not related.

coderbyheart replied:
Unrelated.

MLopezJ reviewed (commented):
.

MLopezJ committed:
fix: use selectos instead of IDs
Use css selectors instead of IDs to set styling

coderbyheart committed:
test(e2e): make movement resolution value higher than timeout

coderbyheart committed:
test(e2e): use current config as test values

coderbyheart committed:
style: remove comment

coderbyheart committed:
fix: use one class per component

coderbyheart committed:
fix: remove unneeded space

coderbyheart committed:
fix: use Bootstrap button for linking

coderbyheart committed:
feat: update texts and rename asset settings to config

MLopezJ commented:
Doing `smoke testing` before `squash and merge`

MLopezJ committed:
style: add space between components
add distance between elements on mobile view

MLopezJ committed:
fix: update field description
from '...Click 'Apply' to apply these settings to the asset.'
to '...Click 'Apply' to upload these settings to the asset.'

Change 'apply' to 'upload'

MLopezJ committed:
fix: update field description
from '...Click 'Apply' to apply these settings to the asset.'
to '...Click 'Apply' to upload these settings to the asset.'

Change 'apply' to 'upload'

MLopezJ committed:
fix: upload form values after click 'Apply' button on Preset component

MLopezJ commented:
Found a bug during `smoke testing`. Link on `Configuration Explainer` is not working. 

Working on resolve it.

coderbyheart committed:
fix: use callbacks in configuration explainer

... so it does not depend on a specific action to perform on click

coderbyheart committed:
fix: fix config form prevent focus from working

coderbyheart committed:
fix: presets remove use of global ids

MLopezJ committed:
docs: remove unused comments

coderbyheart committed:
test(e2e): add utility function to open collabsables

coderbyheart committed:
style: add return type
